README.md
# Storeon websocket

<img src="https://storeon.github.io/storeon/logo.svg" align="right"
     alt="Storeon logo by Anton Lovchikov" width="160" height="142">

Tiny module for [Storeon] which is adding functionality to work with WebSocket. This means that now you can send and receive events by WebSocket.

It is just 401 bytes module (it uses [Size Limit] to control the size) without any dependencies.

[Storeon]: https://github.com/storeon/storeon
[Size Limit]: https://github.com/ai/size-limit

```js
import { websocket } from '@storeon/websocket'

const url = 'ws://localhost:8080'

const store = createStoreon([
  /* all your modules */
  websocket(url)
 ])

// now all dispatched events will be send to server with address ws://localhost:8080
```

![Example of using websocket events functionality](example.gif)

## Installation

```bash
npm install @storeon/websocket
# or
yarn add @storeon/websocket
```

## Usage

This module has two core functionality:

- sending all events to server
- receiving and dispatching events from server

### websocket(url, include, reconnectInterval, pingPongInterval)

First parameter is address of WebSocket server:
```js
type url = String
```

Second parameter is white list of events. 
All events added to this array will be dispatched/sent.

```js
type include = Void | Array<String>
```

If no pass the `include` then all events will be dispatched/sent.

Third parameter is after how many millisecond try reconnect again if connection is lost (default 500):
```js
type reconnectInterval = Void | Number
```

Fourth parameter is how often in millisecond module should check if connection is alive (default 2000):
```js
type pingPongInterval = Void | Number
```

Only `url` is not optional parameter.

### Server

The module is sending event like a string. So before sending it stringify all events.

After receiving message from server the module trying to parse this string to json with `JSON.parse`.
If received message is valid json then module dispatching it in `storeon`.

Each 2000 ms module send `ping` message to server and if server in 2000 ms doesn’t send back message `pong`
then module trying to reconnect to server.

If server close connection or send error then module trying to reconnect to server.

Example implementation of echo server please see in [example](./test/demo/ws_server.js).

## LICENSE

MIT

## Acknowledgments

This module based on [redux-websocket](https://github.com/giantmachines/redux-websocket) repository.



CHANGELOG.md
# Change Log
This project adheres to [Semantic Versioning](http://semver.org/).

## 1.2.0

* Update dependencies

## 1.1.0

* Update dependencies

## 1.0.0

* Update dependencies

## 0.8.0

* Update dependencies

## 0.7.0

* Update dependencies and update code with new lint rules

## 0.6.0

* Update dependencies
* Add test to cover 100% code

## 0.5.1

* Update types for package

## 0.5.0

* Add dual publish to package

## 0.4.0

* Move project to ES2015+. Update dependencies

## 0.3.1

* Remove unnecessary code

## 0.3.0

* Add new options exclude. Refactor into separate param. (by Pavel Prokudin https://github.com/paul-pro)

## 0.2.1

* Fix exporting types (by Pavel Prokudin https://github.com/paul-pro)

## 0.2.0

* Reduce site
* Add types for TypeScript

## 0.1.1

* Add info about server to README

## 0.1

* Initial release.


test/demo/index.js
let useCallback = require('react').useCallback
let Fragment = require('react').Fragment
let render = require('react-dom').render
let h = require('react').createElement
let { StoreContext, connectStoreon } = require('storeon/react')
let { createStoreon } = require('storeon')
let { storeonLogger, storeonDevtools } = require('storeon/devtools')

let { websocket } = require('../../')
let url = 'ws://localhost:8080'

function counter (store) {
  store.on('@init', () => {
    return {
      count1: 0,
      count2: 0
    }
  })
  store.on('inc', state => {
    return {
      count1: state.count1 + 1,
      count2: state.count2 + 1
    }
  })
}

function Tracker (props) {
  let hue = Math.round(255 * Math.random())
  let style = { backgroundColor: 'hsla(' + hue + ', 50%, 50%, 0.2)' }
  return h('div', { className: 'tracker', style }, props.value)
}

function Button (props) {
  let onClick = useCallback(() => {
    props.dispatch(props.event)
  })
  return h('button', { onClick, className: 'action' }, props.text)
}

let Tracker1 = connectStoreon('count1', props => {
  return h(Tracker, {
    value: 'Counter 1: ' + props.count1
  })
})

let Button1 = connectStoreon(props => {
  return h(Button, {
    dispatch: props.dispatch,
    event: 'inc',
    text: 'Increase counter'
  })
})

function App () {
  return h(
    Fragment,
    null,
    h('div', null, 'All dispatched events will be sent to server'),
    h('div', { className: 'container' }, h(Tracker1)),
    h('div', { className: 'buttons' }, h(Button1))
  )
}

let store = createStoreon([
  counter,
  websocket(url),
  storeonLogger,
  storeonDevtools()
])

render(
  h(StoreContext.Provider, { value: store }, h(App)),
  document.querySelector('main')
)


test/index.test.js
let WS = require('jest-websocket-mock')
let { createStoreon } = require('storeon')

let { websocket } = require('../')

let store
let counter
let fakeURL = 'ws://localhost:8080'

beforeEach(() => {
  counter = function (s) {
    s.on('@init', () => {
      return { a: 0, b: 0 }
    })

    s.on('counter/inc', state => {
      return {
        a: state.a + 1,
        b: state.b + 1
      }
    })
    s.on('counter/dec', state => {
      return {
        a: state.a - 1,
        b: state.b - 1
      }
    })
  }
})

afterEach(() => {
  process.env.NODE_ENV = 'development'
})

it('should throw not throw the error in not development mode', () => {
  process.env.NODE_ENV = 'production'
  expect(() => {
    createStoreon(websocket())
  }).toThrow('modules.forEach is not a function')
})

it('should throw the error', () => {
  expect(() => {
    createStoreon(websocket())
  }).toThrow(
    'The url parameter should be a string. ' +
      'For example: "ws://localhost:8080"'
  )
})

it('should send event to server', async () => {
  let server = new WS.WS(fakeURL)
  store = createStoreon([counter, websocket(fakeURL)])
  await server.connected
  store.dispatch('counter/inc')
  await expect(server).toReceiveMessage('{"event":"counter/inc"}')
  server.close()
})

it('should not sending events if server get the error', async () => {
  let server = new WS.WS(fakeURL)
  store = createStoreon([counter, websocket(fakeURL)])
  await server.connected

  store.dispatch('counter/inc')
  await expect(server).toReceiveMessage('{"event":"counter/inc"}')

  server.error()
  store.dispatch('counter/inc')
  expect(store.get()).toEqual({
    a: 2,
    b: 2
  })
})

it('should not sending events if server closes connection', async () => {
  let server = new WS.WS(fakeURL)
  store = createStoreon([counter, websocket(fakeURL)])
  await server.connected

  store.dispatch('counter/inc')
  await expect(server).toReceiveMessage('{"event":"counter/inc"}')

  server.close()
  store.dispatch('counter/inc')
  expect(store.get()).toEqual({
    a: 2,
    b: 2
  })
})

it('should getting event from server', async () => {
  let server = new WS.WS(fakeURL)
  store = createStoreon([counter, websocket(fakeURL)])
  await server.connected
  server.send('{"event":"counter/inc"}')
  expect(store.get()).toEqual({
    a: 1,
    b: 1
  })
  server.close()
})

it('should not change the state if server sends not exist event', async () => {
  let server = new WS.WS(fakeURL)
  store = createStoreon([counter, websocket(fakeURL)])
  await server.connected
  server.send('{"event":"counter/remove"}')
  expect(store.get()).toEqual({
    a: 0,
    b: 0
  })
  server.close()
})

it('should do nothing if server send non valid json string', async () => {
  let server = new WS.WS(fakeURL)
  store = createStoreon([counter, websocket(fakeURL)])
  await server.connected
  server.send('test string')
  expect(store.get()).toEqual({
    a: 0,
    b: 0
  })
  server.close()
})

it('should skip events not listed in include', async () => {
  let server = new WS.WS(fakeURL)
  let include = ['counter/dec']
  store = createStoreon([counter, websocket(fakeURL, { include })])
  await server.connected

  store.dispatch('counter/inc')
  store.dispatch('counter/dec')
  await expect(server).not.toReceiveMessage('{"event":"counter/inc"}')

  server.close()
})

it('should dispatch events only if listed in include', async () => {
  let server = new WS.WS(fakeURL)
  let include = ['counter/inc']
  store = createStoreon([counter, websocket(fakeURL, { include })])
  await server.connected

  server.send('{"event":"counter/inc"}')
  server.send('{"event":"counter/dec"}')
  expect(store.get()).toEqual({
    a: 1,
    b: 1
  })
  server.close()
})

it('should skip events if listed in exclude', async () => {
  let server = new WS.WS(fakeURL)
  let exclude = ['counter/inc']
  store = createStoreon([counter, websocket(fakeURL, { exclude })])
  await server.connected

  store.dispatch('counter/inc')
  store.dispatch('counter/dec')
  await expect(server).not.toReceiveMessage('{"event":"counter/inc"}')

  server.close()
})

it('should dispatch events only if not listed in exclude', async () => {
  let server = new WS.WS(fakeURL)
  let exclude = ['counter/dec']
  store = createStoreon([counter, websocket(fakeURL, { exclude })])
  await server.connected

  server.send('{"event":"counter/inc"}')
  server.send('{"event":"counter/dec"}')
  expect(store.get()).toEqual({
    a: 1,
    b: 1
  })
  server.close()
})

it('should dispatch all received events if include is empty', async () => {
  let server = new WS.WS(fakeURL)
  store = createStoreon([counter, websocket(fakeURL)])
  await server.connected

  server.send('{"event":"counter/inc"}')
  server.send('{"event":"counter/dec"}')
  expect(store.get()).toEqual({
    a: 0,
    b: 0
  })
  server.close()
})

it('should send all received events if include is empty', async () => {
  let server = new WS.WS(fakeURL)
  store = createStoreon([counter, websocket(fakeURL)])
  await server.connected

  store.dispatch('counter/inc')
  store.dispatch('counter/dec')
  await expect(server).toReceiveMessage('{"event":"counter/inc"}')
  await expect(server).toReceiveMessage('{"event":"counter/dec"}')
  server.close()
})

it('should send ping and get back pong', async () => {
  jest.useFakeTimers()
  let server = new WS.WS(fakeURL)
  let send = jest.spyOn(WebSocket.prototype, 'send')
  let listeners = {}
  jest
    .spyOn(WebSocket.prototype, 'addEventListener')
    .mockImplementation((type, callback) => {
      listeners[type] = callback
    })
  store = createStoreon([counter, websocket(fakeURL)])
  jest.runOnlyPendingTimers()
  expect(send).toHaveBeenCalledWith('ping')
  listeners.message({ data: 'pong' })
  jest.runOnlyPendingTimers()
  expect(send).toHaveBeenCalledWith('ping')
  listeners.close()
  server.close()
})

it('should reconnect if not receiving pong', async () => {
  jest.useFakeTimers()
  let mock = jest.spyOn(global, 'WebSocket').mockImplementation(() => {
    return {
      addEventListener: jest.fn(),
      send: jest.fn()
    }
  })
  store = createStoreon([counter, websocket(fakeURL)])
  expect(mock).toHaveBeenCalledTimes(1)
  jest.advanceTimersByTime(2000)
  expect(mock).toHaveBeenCalledTimes(2)
})

it('should reconnect if server send error', async () => {
  jest.useFakeTimers()
  jest.clearAllMocks()
  let listeners = {}
  let mock = jest.spyOn(global, 'WebSocket').mockImplementation(() => {
    return {
      addEventListener: (type, callback) => (listeners[type] = callback),
      send: jest.fn()
    }
  })
  store = createStoreon([counter, websocket(fakeURL)])
  expect(mock).toHaveBeenCalledTimes(1)
  jest.advanceTimersByTime(500)
  listeners.error()
  jest.advanceTimersByTime(500)
  expect(mock).toHaveBeenCalledTimes(2)
})


test/demo/ws_server.js
const WebSocket = require('ws')
const nanoid = require('nanoid')
const chalk = require('chalk')

const wss = new WebSocket.Server({ port: 8080, clientTracking: true })
console.log(
  chalk.yellowBright('Server on port 8080 is listening (press Ctrl+C to exit)')
)
console.log(chalk.whiteBright('ws://localhost:8080'))
console.log('')
wss.on('connection', ws => {
  let id = nanoid(5)
  ws.id = id
  console.log(
    chalk.yellowBright('Client connected. Generate id:'),
    chalk.green(id)
  )
  ws.on('message', data => {
    console.log(chalk.yellowBright('Message: '))
    console.log(chalk.green(data))
    console.log(chalk.yellowBright('from client with id:'), chalk.green(id))
    if (data === 'ping') {
      ws.send('pong')
      return
    }
    // ws.close() // uncomment this line to test recconection
    wss.clients.forEach(client => {
      if (client.readyState === WebSocket.OPEN && client.id !== id) {
        console.log(
          chalk.yellowBright('Send this message to client:'),
          chalk.green(client.id)
        )
        client.send(data)
      }
    })
  })
})


index.js
/**
 *  Storeon is a module to sync events through WebSocket
 *  @param {String} url Address of WebSocket server
 *  @param {Object} [options] Module configuration
 *  @param {String[]} [options.include] List of events to sync
 *  @param {String[]} [options.exclude] List of events to ignore
 *  @param {Number} [options.reconnectInterval] Time (ms) to reconnect
 *  @param {Number} [options.pingPongInterval] Interval (ms) to ping server
 **/
let websocket = function (url, options) {
  options = options || {}
  let include = options.include
  let exclude = options.exclude
  let reconnectInterval = options.reconnectInterval || 500
  let pingPongInterval = options.pingPongInterval || 2000

  if (process.env.NODE_ENV === 'development') {
    if (!url) {
      throw new Error(
        'The url parameter should be a string. ' +
          'For example: "ws://localhost:8080"'
      )
    }
  }

  return function (store) {
    let connection
    let reconnectTimer
    let pingPongTimer
    let pingPongAttempt = false
    let reconnectAttempt = false
    let fromWS = Symbol('ws')
    function reconnect () {
      if (reconnectAttempt) return
      clearTimeout(reconnectTimer)
      reconnectAttempt = true
      reconnectTimer = setTimeout(() => {
        start()
      }, reconnectInterval)
    }
    function message (event) {
      if (event.data === 'pong') {
        pingPongAttempt = false
        return
      }
      try {
        let receive = JSON.parse(event.data)
        receive.value = receive.value || {}
        if (include && !include.includes(receive.event)) return
        if (exclude && exclude.includes(receive.event)) return
        receive.value[fromWS] = true
        store.dispatch(receive.event, receive.value)
      } catch (e) {}
    }

    function pingPong () {
      if (!pingPongAttempt) {
        connection.send('ping')
      } else {
        start()
      }
      pingPongAttempt = !pingPongAttempt
    }

    function start () {
      clearInterval(pingPongTimer)
      clearTimeout(reconnectTimer)
      connection = new WebSocket(url)
      connection.addEventListener('close', reconnect)
      connection.addEventListener('error', reconnect)
      connection.addEventListener('message', message)
      pingPongTimer = setInterval(pingPong, pingPongInterval)
    }
    start()

    store.on('@dispatch', (_, data) => {
      if (
        (data[0] && data[0][0] === '@') ||
        (data[1] && data[1][fromWS]) ||
        connection.readyState !== WebSocket.OPEN ||
        (include && !include.includes(data[0])) ||
        (exclude && exclude.includes(data[0]))
      ) {
        return
      }

      connection.send(
        JSON.stringify({
          event: data[0],
          value: data[1]
        })
      )
    })
  }
}

module.exports = { websocket }
